DataBase = {
    # v.SECC to iMX
    "0x301": {
        "targetContactorsStatus": {
            "start_bit": 0,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" : {
                0: "open",
                1: "closed",                
            }
        },
        "evConnectionState": {
            "start_bit": 1,
            "length": 2,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 3,
            "value_table" : {
                0: "disconnected",
                1: "connected",    
                2: "energyTransferAllowed",   
                3: "error",              
            }
        },
        "chargingState": {
            "start_bit": 3, 
            "length": 3,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 7,
            "value_table" : {
                0: "standby",
                1: "cableCheck",    
                2: "preCharge",   
                3: "charge",
                4: "postCharge",             
            }
        },
        "targetVoltage": {
            "start_bit": 8, 
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": 0,
            "min": 0,
            "max": 6553.5,
            "value_table" : None,
            
        },
        "targetCurrent": {
            "start_bit": 24,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 0.1,
            "offset": 0,
            "min": -3276.8,
            "max": 3276.7,
            "value_table" : None,
        },
        "batteryStateOfCharge": {
            "start_bit": 40, 
            "length": 7,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 100,
            "value_table" : None,
        },
        "cableCheckVoltage": {
            "start_bit": 48, 
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 65535,
            "value_table" : None,
            },        
    },
    "0x308": {
        "vehicleId": {
            "start_bit": 0,
            "length": 48,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 0,
            "value_table" : None,
        },
    },
    "0x30D":{
        "chargeMode": {
            "start_bit": 48,
            "length": 8,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 0,
            "value_table" : {
                0: "unknown",
                1: "scheduled",
                2: "dynamic",
                3: "dynamicBpt"                
            }
        },
        "evMinPower":{
            "start_bit": 32,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 10,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,

        },
        "evMinCurrent":{
            "start_bit": 16,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,
        },
        "evMinVoltage":{
            "start_bit": 0,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,
        }
    },
    "0x30B":{
        "evMaxCurent":{
            "start_bit": 16,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,
        },
        "evMaxPower":{
            "start_bit": 32,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1000,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,
        },
        "evMaxVoltage":{
            "start_bit": 0,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,
        }
    },
    "0x30A":{
        "chargingProfileMaxPowerLimit":{
            "start_bit": 0,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 10,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,
        },
        "timeToFullSoc":{
            "start_bit": 16,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" : None,
        }
    },
    # CharCon to iMX
    "0x5C1":{
        "CHRGR_iAcAct":{
            "start_bit": 4,
            "length": 11,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": -102.2,
            "min": -102.2,
            "max": 102.3,
            "value_table":{
                2046:"Init",
                2047:"Error"
            },
        },
        "CHRGR_iAcAct2":{
            "start_bit": 15,
            "length": 11,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": -102.2,
            "min": -102.2,
            "max": 102.3,
            "value_table" : {
                2046:"Init",
                2047:"Error"
            },
        },
        "CHRGR_iAcAct3":{
            "start_bit": 26,
            "length": 11,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": -102.2,
            "min": -102.2,
            "max": 102.3,
            "value_table" : {
                2046:"Init",
                2047:"Error"
            },
        },
    },
    "0x5C2":{
        "CHRGR_uAcActPha1":{
            "start_bit": 4,
            "length": 12,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.25,
            "offset": 0,
            "min": 0,
            "max": 1023.25,
            "value_table" : {
                4094:"Init",
                4095:"Error"
            }, 
        },
        "CHRGR_uAcActPha2":{
            "start_bit": 16,
            "length": 12,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.25,
            "offset": 0,
            "min": 0,
            "max": 1023.25,
            "value_table" : {
                4094:"Init",
                4095:"Error"
            }, 
        },
        "CHRGR_uAcActPha3":{
            "start_bit": 28,
            "length": 12,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.25,
            "offset": 0,
            "min": 0,
            "max": 1023.25,
            "value_table" : {
                4094:"Init",
                4095:"Error"
            },
        },
    },
    "0x5C0":{
        "CHRGR_iDcAct":{
            "start_bit": 20,
            "length": 13,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": -409.4,
            "min": -409.4,
            "max": 409.5,
            "value_table" :{
                8190:"Init",
                8191:"Error"
            },
        },
        "CHRGR_stOperAct":{
            "start_bit": 4,
            "length": 4,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 15,
            "value_table" :{
                0:"Init",
                1:"Standby_noAC",
                2:"standby_ready2charge",
                3:"charging",
                4:"reserved",
                5:"error_recoverable",
                6:"error_nonrecoverable",
                7:"unused",
                8:"unused",
                9:"unused",
                10:"unused",
                11:"unused",
                12:"unused",
                13:"unused",
                14:"unused",
                15:"shutdown"
            }, 
        },
        "CHRGR_uDcAct":{
            "start_bit": 8,
            "length": 12,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.25,
            "offset": 0,
            "min": 0,
            "max": 1023.25,
            "value_table" :{
                4094:"Init",
                4095:"Error"
            },
        },
    },
    "0x5C3":{
        "CHRGR_stErr":{
            "start_bit": 32,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"no warning",
                1:"warning present",
            },
        },
        "CHRGR_stHvIntlck":{
            "start_bit": 33,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"open",
                1:"closed",
            },
        },
        "DC1_bReqMIL":{
            "start_bit": 6,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"MIL Off",
                1:"MIL On",
            },
        },
        "DC1_stT30c":{
            "start_bit": 4,
            "length": 2,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 3,
            "value_table" :{
                0:"T30c Off",
                1:"T30c On",
                2:"init",
                3:"error"
            },
        },
    },
    "0x5D0":{
        "DC1_Temp":{
            "start_bit": 39,
            "length": 8,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": -40,
            "min": -40,
            "max": 213,
            "value_table" :{
                0xFE:"Init",
                0xFF:"Error",
            },
        },
    },
    # iMX to CharCon
    "0x186":{
        "mc_VCU_OBCControl":{
            "start_bit": 0,
            "length": 4,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 15,
            "value_table" :None,
        },
        "VCU_OBCMode_Cmd":{
            "start_bit": 4,
            "length": 4,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 15,
            "value_table" :{
                0:"Init",
                1:"reserved",
                2:"standby",
                3:"charging",
                4:"reserved",
                5:"reserved",
                6:"reserved",
                7:"unused",
                8:"unused",
                9:"unused",
                10:"unused",
                11:"unused",
                12:"unused",
                13:"unused",
                14:"unused",
                15:"shutdown",
            }, 
        },
        "VCU_uDcReqChrgr":{
            "start_bit": 8,
            "length": 12,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.25,
            "offset": 0,
            "min": 0,
            "max": 1023.25,
            "value_table" :{
                4094:"Init",
            }, 
        },
        "VCU_iDcLimChrgr":{
            "start_bit": 20,
            "length": 13,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": -409.4,
            "min": -409.4,
            "max": 409.5,
            "value_table" :{
                8190:"Init",
            },
        },
        "VCU_iAcMaxAllwdChrgr":{
            "start_bit": 33,
            "length": 11,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": -102.2,
            "min": -102.2,
            "max": 102.3,
            "value_table" :{
                2046:"Init",
                2047:"Error",
            },
        },
        "CRC_VCU_OBCControl":{
            "start_bit": 56,
            "length": 8,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 255,
            "value_table" :None,
        },
    },
    
    "0x102":{
        "CrashDetect":{
            "start_bit": 0,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"No Crash",
                1:"Crash",
            },
        },
        "CrashConfirmed_Sts":{
            "start_bit": 2,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"Not Active",
                1:"Active",
            }, 
        },
        "mc_ACU_Crash":{
            "start_bit": 3,
            "length": 4,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 15,
            "value_table" :None,
        },
        "CRC_VCU_OBCControl":{
            "start_bit": 8,
            "length": 8,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 255,
            "value_table" :None,
        },
    },
    
    "0x630":{
        "NWM_WkUp_BCU_Rx":{
            "start_bit": 0,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"Default",
                1:"WakeUp",
            },
        },
        "NWM_WkUp_BCU_Reason":{
            "start_bit": 1,
            "length": 5,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 31,
            "value_table" :{
                0:"Network",
                # There are other states which are ignored as only "Network" is important to us
            }, 
        },
        "NWM_WkUp_BCU_NodeAddress":{
            "start_bit": 8,
            "length": 5,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 63,
            "value_table" :None,
        },
        "NWM_Sleep_BCU_Req":{
            "start_bit": 15,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"Default",
                1:"Sleep",
            },
        },
    },
    
    "0x182":{
        "mc_VCU_DCDCControl":{
            "start_bit": 0,
            "length": 4,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 15,
            "value_table" :None,
        },
        "VCU_VoltageDemandHV":{
            "start_bit": 4,
            "length": 10,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1021,
            "value_table" :{
                0x3FE:"Init",
                0x3FF:"Error"
            }, 
        },
        "VCU_CurrentLimitHV":{
            "start_bit": 14,
            "length": 11,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.0625,
            "offset": -64,
            "min": -64,
            "max": 63.8125,
            "value_table" :{
                0x7FE:"Init",
                0x7FF:"Error"
            }, 
        },
        "VCU_CurrentLimitLV":{
            "start_bit": 25,
            "length": 11,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": -1023,
            "min": -1023,
            "max": 1022,
            "value_table" :{
                0x7FE:"Init",
                0x7FF:"Error"
            },
        },
        "VCU_VoltageDemandLV":{
            "start_bit": 36,
            "length": 9,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.125,
            "offset": 0,
            "min": 0,
            "max": 63.625,
            "value_table" :{
                0x1FE:"Init",
                0x1FF:"Error"
            },
        },
        "VCU_DCDCMode_Cmd":{
            "start_bit": 45,
            "length": 3,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 7,
            "value_table" :{
                0x0:"Standby",
                0x1:"Buck",
                0x2:"Boost",
                0x3:"Reserved",
            },
        },
        "VCU_stHvDeactvnDC":{
            "start_bit": 50,
            "length": 2,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 3,
            "value_table" :{
                0x0:"no deactivation",
                0x1:"HV deactivation",
                0x2:"reserved",
                0x3:"reserved",
            },
        },
        "CRC_VCU_DCDCControl":{
            "start_bit": 56,
            "length": 8,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 255,
            "value_table" :None,
        },
    },
    
    "0x188":{
        "mc_VCU_SysControl":{
            "start_bit": 0,
            "length": 4,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 15,
            "value_table" :None,
        },
        "KL15_VCU_Sts":{
            "start_bit": 4,
            "length": 3,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 7,
            "value_table" :{
                0x0:"OFF",
                0x1:"ON",
                0x2:"Fault on relay contact",
                0x3:"Fault on relay coil",
                0x4:"Feedback Fault",
                0x5:"Reserved",
                0x6:"Reserved",
                0x7:"Reserved",
            }, 
        },
        "HV_HVIL_Sts":{
            "start_bit": 24,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0x0:"HVIL Open",
                0x1:"HVIL Closed"
            }, 
        },
        "CRC_VCU_SysControl":{
            "start_bit": 56,
            "length": 8,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 255,
            "value_table" :None,
        },
    },
    
    # iMX to v.SECC
    "0x304":{
        "limitVoltageMin":{
            "start_bit": 0,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": 0,
            "min": 0,
            "max": 6553.5,
            "value_table" :None,
        },
        "limitVoltageMax":{
            "start_bit": 16,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": 0,
            "min": 0,
            "max": 6553.5,
            "value_table" :None,
        },
        "limitPowerMax":{
            "start_bit": 32,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 10,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" :None,
        },
         "limitPowerMin":{
            "start_bit": 48,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 10,
            "offset": 0,
            "min": 0,
            "max": 655350,
            "value_table" :None,
        },
    },
    "0x305":{
        "limitCurrentMin":{
            "start_bit": 0,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 0.1,
            "offset": 0,
            "min": -3276.8,
            "max": 3276.7,
            "value_table" :None,
        },
        "limitCurrentMax":{
            "start_bit": 16,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 0.1,
            "offset": 0,
            "min": -3276.8,
            "max": 3276.7,
            "value_table" :None,
        },
    },
    "0x302":{
        "contactorsStatus":{
            "start_bit": 0,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"open",
                1:"closed",
            }, 
        },
        "operationalStatus":{
            "start_bit": 1,
            "length": 1,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 1,
            "value_table" :{
                0:"operative",
                1:"inoperative",
            },
        },
        "isolationStatus":{
            "start_bit": 2,
            "length": 3,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 1,
            "offset": 0,
            "min": 0,
            "max": 7,
            "value_table" :{
                0:"invalid",
                1:"valid",
                2:"warning",
                3:"fault",
            },
        },
        "drivenVoltage":{
            "start_bit": 8,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": 0,
            "min": 0,
            "max": 6553.5,
            "value_table" :None, 
        },
        "drivenCurrent":{
            "start_bit": 24,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 0.1,
            "offset": 0,
            "min": -3276.8,
            "max": 3276.7,
            "value_table" :None, 
        },
        "temperature":{
            "start_bit": 40,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 0.1,
            "offset": 0,
            "min": -3276.8,
            "max": 3276.7,
            "value_table" :None, 
        },
    },
    "0x303":{
        "measuredVoltage":{
            "start_bit": 0,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'unsigned',
            "factor": 0.1,
            "offset": 0,
            "min": 0,
            "max": 6553.5,
            "value_table" :None,
        },
        "measuredCurrent":{
            "start_bit": 16,
            "length": 16,
            "byte_order": 'intel',
            "value_type": 'signed',
            "factor": 0.1,
            "offset": 0,
            "min": -3276.8,
            "max": 3276.7,
            "value_table" :None,
        },
    },
}